src/models/losses.py

```diff
@@ -4,6 +4,327 @@
 import torch
 from torch import nn
 
+# pylint: disable=too-many-arguments
+# pylint: disable=too-many-locals
+
+
+def generator_loss(
+    logits: dict[str, dict[str, torch.Tensor]],
+    inception_encoder: Any,
+    fake_imgs: torch.Tensor,
+    real_labels: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    vgg_encoder: Any,
+    real_imgs: torch.Tensor,
+    device: Any,
+    const_dict: dict[Any, Any],
+) -> Any:
+    """Calculate the loss for the generator.
+
+    Args:
+        logits: Dictionary with fake/real and word-level/uncond/cond logits
+
+        inception_encoder: The inception encoder model.
+
+        fake_imgs: The fake images generated by the generator. [shape: (batch_size, 3, 256, 256)]
+
+        real_labels: Label for "real" image as predicted by discriminator,
+        this is a tensor of ones. [shape: (batch_size, 1)].
+
+        words_emb: The embeddings for all the words in the captions.
+        shape: (batch_size, embedding_size, max_caption_length)
+
+        sent_emb: The embeddings for the sentences.
+        shape: (batch_size, embedding_size)
+
+        match_labels: Tensor of shape: (batch_size, 1).
+        This is of the form torch.tensor([0, 1, 2, ..., batch-1])
+
+        cap_lens: The length of the 'actual' captions in the batch [without padding]
+        shape: (batch_size, 1)
+
+        class_ids: The class ids for the instance. shape: (batch_size, 1)
+
+        vgg_encoder: The VGG encoder model.
+
+        real_imgs: The Real Images from the dataset. shape: (batch_size, 3, 256, 256)
+
+        device: The device to run the model on.
```

Comment:
device is being used to transfer the mask tensor created inside DAMSM loss into GPU. Since this tensor is created inside this loss for purpose of this loss, might have to pass in device if we want that tensor to be in GPU, if not can get rid of the device argument but then the tensor operation will be done in CPU which might be slow. So what do you think.

Reply:
usually losses are supposed to be purely functional and there's almost no difference between running on GPU or CPU so loss is usually made unaware of device

if operations that involve `mask` operate on big tensors and can be a potential bottleneck then ok, it makes sense to make loss aware of device and do calculations on GPU sometimes

but if you think that those operations don't contribute much to the overall time, then I would suggest to remove the device dependancy and calculate everything on CPU

Reply:
i see, the tensor is (batch, batch) and we use the operation twice once for masking local damsm features other for global.
for now i have removed device from the loss, so i guess you can verify and resolve.